glommio/src/executor/mod.rs

```diff
@@ -2091,30 +2110,17 @@ impl ExecutorProxy {
     ///
     /// [`RefMut`]: https://doc.rust-lang.org/std/cell/struct.RefMut.html
     #[inline(always)]
-    // FIXME: This is a bit less efficient than it needs, because the scoped thread
-    // local key does lazy initialization. Every time we call into this, we are
-    // paying to test if this is initialized. This is what I got from objdump:
-    //
-    // 0:    50                      push   %rax
-    // 1:    ff 15 00 00 00 00       callq  *0x0(%rip)
-    // 7:    48 85 c0                test   %rax,%rax
-    // a:    74 17                   je     23  <== will call into the
-    // initialization routine c:    48 8b 88 38 03 00 00    mov
-    // 0x338(%rax),%rcx <== address of the head 13:   48 8b 80 40 03 00 00
-    // mov    0x340(%rax),%rax <== address of the tail 1a:   8b 00
-    // mov    (%rax),%eax 1c:   3b 01                   cmp    (%rcx),%eax <==
-    // need preempt 1e:   0f 95 c0                setne  %al
-    // 21:   59                      pop    %rcx
-    // 22:   c3                      retq
-    // 23    <== initialization stuff
-    //
-    // Rust has a thread local feature that is under experimental so we can maybe
-    // switch to that someday.
-    //
-    // We will prefer to use the stable compiler and pay that unfortunate price for
-    // now.
     pub fn need_preempt(&self) -> bool {
-        LOCAL_EX.with(|local_ex| local_ex.need_preempt())
+        #[cfg(not(feature = "native-tls"))]
+        return LOCAL_EX.with(|local_ex| local_ex.need_preempt());
+
+        #[cfg(feature = "native-tls")]
+        return unsafe {
+            LOCAL_EX
+                .as_ref()
+                .expect("this thread doesn't have a LocalExecutor running")
```

Comment:
I'd be interested in seeing what kind of code we generate.
There's no need to panic here if the pointer is null, just return false.
That could likely lead to a more efficient implementation since there's no need to inline the panic handler

Reply:
OIC; let me do that. 

Using std-tls:
```
glommio::executor::ExecutorProxy::need_preempt:
 push    rax
 lea     rdi, [rip, +, _ZN7glommio8executor8LOCAL_EX3FOO7__getit5__KEY17h9afbb68dc0c48f46E@TLSLD]
 call    __tls_get_addr
 cmp     dword, ptr, [rax, +, _ZN7glommio8executor8LOCAL_EX3FOO7__getit5__KEY17h9afbb68dc0c48f46E@DTPOFF], 1
 jne     .LBB640_2
 lea     rax, [rax, +, _ZN7glommio8executor8LOCAL_EX3FOO7__getit5__KEY17h9afbb68dc0c48f46E@DTPOFF+8]
 jmp     .LBB640_3
.LBB640_2:
 call    std::thread::local::fast::Key<T>::try_initialize
 test    rax, rax
 je      .LBB640_5
.LBB640_3:
 mov     rax, qword, ptr, [rax]
 test    rax, rax
 je      .LBB640_6
 mov     rax, qword, ptr, [rax, +, 24]
 mov     rcx, qword, ptr, [rax, +, 2600]
 mov     rax, qword, ptr, [rax, +, 2608]
 mov     eax, dword, ptr, [rax]
 cmp     eax, dword, ptr, [rcx]
 setne   al
 pop     rcx
 ret
.LBB640_6:
 lea     rdi, [rip, +, .L__unnamed_92]
 lea     rdx, [rip, +, .L__unnamed_93]
 mov     esi, 72
 call    std::panicking::begin_panic
 ud2
.LBB640_5:
 lea     rdi, [rip, +, .L__unnamed_98]
 lea     rcx, [rip, +, .L__unnamed_99]
 lea     r8, [rip, +, .L__unnamed_100]
 mov     rdx, rsp
 mov     esi, 70
 call    qword, ptr, [rip, +, _ZN4core6result13unwrap_failed17hf836a820f86fa4a9E@GOTPCREL]
 ud2
```

Using native-tls (as in the PR):
```
glommio::executor::ExecutorProxy::need_preempt:
 push    rax
 data16
 lea     rdi, [rip, +, _ZN7glommio8executor8LOCAL_EX17ha0f5da69d3f777eaE@TLSGD]
 data16
 data16
 rex64
 call    __tls_get_addr
 mov     rax, qword, ptr, [rax]
 test    rax, rax
 je      .LBB640_1
 mov     rax, qword, ptr, [rax, +, 24]
 mov     rcx, qword, ptr, [rax, +, 2600]
 mov     rax, qword, ptr, [rax, +, 2608]
 mov     eax, dword, ptr, [rax]
 cmp     eax, dword, ptr, [rcx]
 setne   al
 pop     rcx
 ret
.LBB640_1:
 lea     rdi, [rip, +, .L__unnamed_92]
 lea     rdx, [rip, +, .L__unnamed_639]
 mov     esi, 48
 call    qword, ptr, [rip, +, _ZN4core6option13expect_failed17h7e5fdafed144cc7fE@GOTPCREL]
 ud2
```

Using native-tls and your suggestion to return `false` instead of panicking
```
glommio::executor::ExecutorProxy::need_preempt:
 push    rax
 data16
 lea     rdi, [rip, +, _ZN7glommio8executor8LOCAL_EX17ha0f5da69d3f777eaE@TLSGD]
 data16
 data16
 rex64
 call    __tls_get_addr
 mov     rax, qword, ptr, [rax]
 test    rax, rax
 je      .LBB640_1
 mov     rax, qword, ptr, [rax, +, 24]
 mov     rcx, qword, ptr, [rax, +, 2600]
 mov     rax, qword, ptr, [rax, +, 2608]
 mov     eax, dword, ptr, [rax]
 cmp     eax, dword, ptr, [rcx]
 setne   cl
 jmp     .LBB640_3
.LBB640_1:
 mov     cl, 2
.LBB640_3:
 cmp     cl, 2
 setne   al
 and     al, cl
 pop     rcx
 ret
```